dll/win32/winmm/playsound.c

```diff
@@ -39,6 +40,50 @@ typedef struct tagWINE_PLAYSOUND
 static WINE_PLAYSOUND *PlaySoundCurrent;
 static BOOL bPlaySoundStop;
 
+/* An impersonation-aware equivalent of ExpandEnvironmentStringsW */
+static DWORD PlaySound_ExpandEnvironmentStrings(LPCWSTR lpSrc, LPWSTR lpDst, DWORD nSize)
+{
+    HANDLE hToken;
+    WCHAR szExpanded[1024]; /* ExpandEnvironmentStringsForUser doesn't support retrieving buffer size */
+    DWORD dwLength = 0;
+
+    if (!OpenThreadToken(GetCurrentThread(),
+                         TOKEN_QUERY | TOKEN_IMPERSONATE | TOKEN_DUPLICATE,
+                         TRUE,
+                         &hToken))
+    {
+        DWORD dwError = GetLastError();
+
+        if (dwError == ERROR_NO_TOKEN)
+        {
+            /* We are not impersonating, forward this to ExpandEnvironmentStrings */
+            return ExpandEnvironmentStringsW(lpSrc, lpDst, nSize);
+        }
+
+        ERR("OpenThreadToken failed (0x%x)\n", dwError);
+        return 0;
+    }
+
+    if (!ExpandEnvironmentStringsForUserW(hToken, lpSrc, szExpanded, ARRAY_SIZE(szExpanded)))
```

Comment:
Why not just using directly the lpDst and nSize parameters given to the PlaySound_ExpandEnvironmentStrings ?

Reply:
Fixed.